app/models/log_ticket.rb

```diff
@@ -0,0 +1,27 @@
+class LogTicket < ActiveRecord::Base
+  enum backend: [ :s3, :local ]
+
+  scope :latest_pending, -> { limit(1).lock(true).select("id").where(status: "pending") }
+
+  def self.pop(key = nil, directory = nil)
+    scope = latest_pending
+    scope = scope.where(key: key) if key
+    scope = scope.where(directory: directory) if directory
+    sql = scope.to_sql
+
+    find_by_sql(["UPDATE #{quoted_table_name} SET status = ? WHERE id IN (#{sql}) RETURNING *", 'processing']).first
+  end
+
+  def filesystem
+    @fs ||=
+      if s3?
```

Comment:
IMHO it will be better to hide this branching and passing credentials in `RubygemFs` class since it is really similar code to that used in [RubygemFs#instance](https://github.com/rubygems/rubygems.org/blob/e00090001e1b8e6d2d794a27dec9cbce720c5dfe/lib/rubygem_fs.rb#L2).

for example:

```ruby
@filesystem ||= RubygemFS.new(directory: directory, s3: s3?)
```